### PR TITLE
TRT-2848: Refresh summary tables incrementally during prow load

### PR DESCRIPTION
**Pull request by mstaeble**
## Summary

- Refreshes `test_daily_totals` and `test_cumulative_summaries` incrementally as part of each prow load batch, replacing the separate full-refresh step that re-scanned the entire `prow_job_run_tests` table
- Extracts batch writing logic from prow.go into a dedicated `pgwriter` package with focused, single-responsibility functions for each SQL step
- Parallelizes carry-forward across releases (4 workers via `errgroup.Group`) to reduce carry-forward time from ~3 minutes to ~50 seconds
- Pins `currentDate` at load start and threads it through all SQL to prevent midnight-crossing corruption
- Adds 32 integration tests covering job runs, annotations, PRs, test results, daily/cumulative summaries, carry-forward, timestamp tracking, and edge cases

## Staging validation

Tested against the staging database with multiple lookback windows:

- **5-minute lookback**: ~69 prow jobs, completed in under 2 minutes including summary upserts
- **1-hour lookback**: 17-23 jobs with 30-43K tests, completed in ~1 minute steady-state
- **12-hour lookback (cold start simulation)**: 364 candidate jobs, completed successfully with correct daily totals and cumulative summaries
- **Carry-forward parallelization**: reduced from ~3 minutes (sequential) to ~50 seconds (4 workers) on staging

## Test plan

- [x] `make lint` passes (0 issues)
- [x] `go test ./pkg/...` passes
- [x] `make integration` passes (150 tests)
- [x] Verified against staging with multiple lookback windows (5m, 1h, 12h)
- [ ] Compare daily totals and cumulative summaries against full-refresh baseline

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## New Features
- Improved persistence of job runs, annotations, pull requests, test results, and test outputs.
- Daily and cumulative summaries now remain consistent across batches, releases, suites, and missed dates.
- Historical summaries automatically carry forward to fill eligible gaps.
- Added explicit date-range backfilling for daily and cumulative summaries.
- Pull-request metadata is preserved when records are reprocessed.

## Bug Fixes
- Improved handling of duplicates, soft-deleted records, empty batches, and invalid future-dated results.
- Refresh operations now focus on materialized views while summary updates occur during data loading.

### CODE REVIEW

File: pkg/dataloader/prowloader/pgwriter/pgwriter.go [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+		for day := minDate; !day.After(tomorrow); day = day.AddDays(1) {
+			if err := ensureCumulativeSummaryRows(ctx, tx, day, release); err != nil {
+				return err
+			}
+			if err := updateCumulativeSummaries(ctx, tx, day, release); err != nil {
+				return err
+			}
+		}
+	}
+	log.WithField("elapsed", time.Since(stepStart)).Debug("upserted test_cumulative_summaries")
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/dataloader/prowloader/pgwriter/pgwriter.go:425]: Added a guard in `upsertSummaryTables` that fails the batch if any test results are dated after tomorrow. Also added `TestFutureDatedTestResultsRejectBatch` to verify the transaction rolls back.

File: pkg/dataloader/prowloader/pgwriter/pgwriter.go [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+	latestDate, err := findLatestDateWithData(ctx, dbc, release, tomorrow)
+	if err != nil {
+		return fmt.Errorf("finding latest date for release %s: %w", release, err)
+	}
+
+	nextDay := latestDate.AddDays(1)
+	if nextDay.After(tomorrow) {
+		log.WithField("release", release).Debug("cumulative summaries already up to date through tomorrow")
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/dataloader/prowloader/pgwriter/pgwriter.go:624]: We should not have a state where there is no data for releases. We carry everything forward, even inactive releases. For the database to be in this state would be catastrophic and should not allow the rest of the loading to proceed.

File: pkg/dataloader/prowloader/pgwriter/pgwriter.go [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+					AND dt.prow_job_id = bd.prow_job_id
+					AND dt.suite_id = bd.suite_id
+					AND dt.release = $2
+					AND dt.date = $1
+			)
+	`, day, release); err != nil {
+		return fmt.Errorf("ensuring test_daily_totals rows for release %s date %s: %w", release, day, err)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/dataloader/prowloader/pgwriter/pgwriter.go:500]: This is a false finding. `civil.Date` works correctly as a pgx bind parameter. pgx serializes it via its `String()` method, which returns YYYY-MM-DD format, and PostgreSQL implicitly casts the text to date. This is validated by every integration test in this PR that writes or queries by date. No conversion to `time.Time` is needed.

**Review by petr-muller — APPROVED** [review]
I have reviewed the code and had some convos with Claude Code about and found no problems, but I do not feel too confident about knowing this part of Sippy that well, so LGTM+hold for the case you want someone better with DBs than me to look as well. Unhold as needed

/hold

File: pkg/dataloader/prowloader/prow.go [hunk 4]
```diff
@@ -127,9 +124,8 @@ func New(
 		ghCommenter:                  ghCommenter,
 		promPusher:                   promPusher,
 		loadSince:                    loadSince,
+		currentDate:                  civil.DateOf(time.Now().UTC()),
```
Comment (neisw) [inline pkg/dataloader/prowloader/prow.go:127]: This could be ok for now but what are the implications when we get to loading historical data?  There is more to it than just this so it is really just an observation for now, nothing to take action on

Comment (neisw) [reply pkg/dataloader/prowloader/prow.go:127]: Or maybe this works as any historical updates will need to be brought all the way forward to today.

File: pkg/sippyserver/server.go [hunk 5]
```diff
@@ -390,24 +384,11 @@ type RefreshOptions struct {
 	RefreshOnlyIfEmpty bool
 }
 
-// RefreshData runs the normal incremental refresh of all summary tables
-// and materialized views. Used by the load command and the serve loop.
+// RefreshData refreshes materialized views. Summary tables
+// (test_daily_totals, test_cumulative_summaries) are updated
+// incrementally by the prow loader; use "sippy backfill" to repair them.
 func RefreshData(dbc *db.DB, cacheClient cache.Cache, opts RefreshOptions) error {
 	log.Infof("Refreshing data")
```
Comment (neisw) [inline pkg/sippyserver/server.go:391]: e2e seems to be passing so seed-data may not be an issue.  Without new data coming in running the refresh command will just be refreshing the materialized views so all may be fine, posting the review from ai-helpers in this area just to confirm there are no real concerns raised here:

 [High] LSP: RefreshData() behavioral change -- server.go: Removing dailysummary.Refresh and cumulativesummary.Refresh from RefreshData() silently changes the function's contract for all
  callers, not just the prow loader:

  - cmd/sippy/load.go -- calls RefreshData after prow load. Since the prow loader now handles summaries incrementally, this caller gets equivalent behavior. OK.
  - cmd/sippy/refresh.go -- explicit "refresh" command. After this PR, sippy refresh will not update daily totals or cumulative summaries. This is a breaking change for operators.
  - cmd/sippy/seed_data.go -- seeds data then refreshes. Summary tables will not be rebuilt after seeding.

  The PR description mentions "use sippy backfill to repair them," but the refresh and seed-data commands' help text and behavior are not updated to reflect this change. Recommendation:
  Either (a) have those commands also invoke backfill/carry-forward for summary tables, or (b) explicitly document the behavioral change in command help text and the PR description.

File: pkg/dataloader/prowloader/prow.go (hunk 4 above)
Comment (mstaeble) [reply pkg/dataloader/prowloader/prow.go:127]: When we get to loading historical data, we will want a more deliberate command for that. Thinking aloud, we'll want to load the oldest data first, loading a few days at a time. Then we'll want to separately backfill the summary tables. The commands already exist for that summary backfill that go deliberately day by day.

File: pkg/sippyserver/server.go (hunk 5 above)
Comment (mstaeble) [reply pkg/sippyserver/server.go:391]: The change to "sippy refresh" is deliberate. The backfill command is the tool to use to re-calculate summary tables.

There is probably a gap in the seed_data command. I had not considered that.

Comment (mstaeble) [reply pkg/sippyserver/server.go:391]: For the seed data path, it has been converted to use the `BackfillData` function already. It should be functioning properly.
